source/opt/control_dependence.h

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2021 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_OPT_CONTROL_DEPENDENCE_H_
+#define SOURCE_OPT_CONTROL_DEPENDENCE_H_
+
+#include <algorithm>
+#include <cstdint>
+#include <functional>
+#include <map>
+#include <vector>
+
+#include "source/opt/cfg.h"
+#include "source/opt/dominator_analysis.h"
+
+namespace spvtools {
+namespace opt {
+
+struct ControlDependence {
+  enum class DependenceType {
+    kConditionalBranch,
+    kSwitchCase,
+    kEntry,
+  };
+  // The label of the source of this dependence, i.e. the dependee.
+  uint32_t source;
+  // The label of the target of this dependence, i.e. the dependent.
+  uint32_t target;
+  // The type of dependence: either a conditional branch or a switch-case
+  // (either a numbered case or default).
+  DependenceType dependence_type;
+  // The label for the value which this dependence is on.
+  // For conditional branches, this is the branch condition, and
+  // for switch cases, this is the value on which the switch is performed.
+  uint32_t dependent_value_label;
+  // For switch cases, the values of the cases for this dependence.
+  std::vector<uint32_t> switch_case_values;
+  // For switch cases, true if this dependence happens when the default branch
+  // is taken.
+  bool is_switch_default;
+  // For conditional branches, the value of the condition required for the
+  // dependence to happen.
+  bool condition_value;
```

Comment:
That makes things (IMO significantly) more complicated on the caller side. Besides the redundant variables for other types (which can't really be resolved cleanly in C++ to my knowledge), is the something in particular you don't you like about the current layout?

Reply:
The problem with have the three different value is that all code that uses it will have to have three cases that are similar, but different.  We can talk about this offline, but it would be best to see how it will be used.  This will make it easier to develop a better interface for this class/struct.